pkg/reconciler/namespace/namespace.go

```diff
@@ -0,0 +1,165 @@
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Don't modify the informers copy
+	newNamespace := original.DeepCopy()
+
+	err = c.reconcile(ctx, newNamespace)
+	if err != nil {
+		c.Recorder.Event(newNamespace, corev1.EventTypeWarning, "InternalError", err.Error())
+	}
+	return err
+}
+
+func (c *Reconciler) reconcile(ctx context.Context, ns *corev1.Namespace) error {
+	cfg := config.FromContext(ctx)
+
+	// Only create wildcard certs for the default domain
+	defaultDomain := cfg.Domain.LookupDomainForLabels(nil)
+	if defaultDomain == "" {
+		c.Logger.Infof("Skipping wildcard certificate creation for namespace %s. No default domain defined.", ns.Name)
+		return nil
+	}
+
+	domain, err := wildcardDomain(cfg.Network.DomainTemplate, defaultDomain, ns.Name)
+	if err != nil {
+		c.Logger.Errorf("Failed to apply domain template %s to domain %s and namespace %s: %v", cfg.Network.DomainTemplate, defaultDomain, ns.Name, err)
+		return err
+	}
+
+	desiredCert := makeDesiredCertificate(ns, domain)
+	existingCert, err := c.knCertificateLister.Certificates(ns.Name).Get(ns.Name)
+	if apierrs.IsNotFound(err) {
+		cert, err := c.ServingClientSet.NetworkingV1alpha1().Certificates(ns.Name).Create(desiredCert)
+		if err != nil {
+			c.Logger.Errorw("Failed to create namespace certificate", zap.Error(err))
+			c.Recorder.Eventf(ns, corev1.EventTypeWarning, "CreationFailed",
+				"Failed to create Knative certificate %s/%s: %v", ns.Name, desiredCert.ObjectMeta.Name, err)
+			return err
+		}
+		c.Recorder.Eventf(cert, corev1.EventTypeNormal, "Created",
+			"Created Knative Certificate %s/%s", ns.Name, cert.ObjectMeta.Name)
+	} else if err != nil {
+		return err
+	} else if !equality.Semantic.DeepEqual(existingCert.Spec, desiredCert.Spec) {
+		copy := existingCert.DeepCopy()
+		copy.Spec = desiredCert.Spec
+		_, err := c.ServingClientSet.NetworkingV1alpha1().Certificates(copy.Namespace).Update(copy)
+		if err != nil {
+			c.Logger.Errorw("Failed to update Knative Certificate", zap.Error(err))
+			c.Recorder.Eventf(existingCert, corev1.EventTypeWarning, "UpdateFailed",
+				"Failed to update Knative Certificate %s/%s: %v", desiredCert.Namespace, desiredCert.Name, err)
+			return err
+		}
+		c.Recorder.Eventf(existingCert, corev1.EventTypeNormal, "Updated",
+			"Updated Spec for Knative Certificate %s/%s", desiredCert.Namespace, desiredCert.Name)
+		return nil
+	}
+
+	return nil
+}
+
+func makeDesiredCertificate(namespace *corev1.Namespace, dnsName string) *v1alpha1.Certificate {
+	return &v1alpha1.Certificate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      namespace.Name,
+			Namespace: namespace.Name,
+			Labels: map[string]string{
+				networking.OwnerNamespaceLabelKey: namespace.Name,
+			},
+		},
+		Spec: v1alpha1.CertificateSpec{
+			DNSNames:   []string{dnsName},
+			SecretName: namespace.Name,
+		},
+	}
+}
+
+func wildcardDomain(tmpl, domain, namespace string) (string, error) {
```

Comment:
This takes a strong dependency on the default DomainTemplate 

> {{.Name}}.{{.Namespace}}.{{.Domain}}

As this is user configurable, it will break if user changes this template. where namespace is not the primary domain

We also recently added support for configuring the domain template based on annotations. [4210](https://github.com/knative/serving/pull/4210/files).

Reply:
Yes, this is expected.  DomainTemplate is a fairly advanced use case, and my expectation is that in that advanced use case you'd effectively disable this controller.

Reply:
(or frankly, roll your own version)

Reply:
That's fine the point i want to make is Annotations needs to be include in the DomainTemplateValues{} as this is now exposed as part of DomainTemplate. The wildcard is for the service name, rest of the url should follow the same domainTemplate which includes values namespace,Annotations, domain.

Reply:
In this case the annotations would come from the ObjectMeta of the namespace?

Reply:
Seems like the Annotation is passed from Route (https://github.com/knative/serving/blob/master/pkg/reconciler/route/route.go#L221), which means it could vary btw Routes within  a namespace. If the Annotation differs btw Routes, there will be no wildcard cert that could be used across Routes within the namespace.

I would say let's start with:
```
type DomainTemplateValues struct {
Name string
Namespace string
Domain string
}
```
If user wants to add Annotation to Route to configure host name per Route, then they can disable the wildcard cert by whitelisting namespace.

About how to support Annotation, I think we may need more discussions.

Reply:
I've just left it as-is for now. It seems reasonable to expect people to not opt-in to namespace certs if they want to also add route annotations.

Reply:
I have two issues with this:
1. This code is replicating the domain name construction logic from domains.go - DomainNameFromTemplate
2. Its copy is subtly different for no apparent reason (aka it doesnt include annotations)

I think the logic should be: 
1. create or get the actual domain name (get URL from status or call or replicate DomainNameFromTemplate)
2. replace everything to the left of the first "." with a *

So basically:
foo.bar.com becomes *.bar.com

I think having two copies of DomainNameFromTemplate logic that are different is a bug that should be fixed.

Note that domains.go also uses annotations from the route and its on the operator (who modified the domain template to use annotations) to make sure the result is valid and consistent. 

I dont see why the operator should have to reimplement the wonderful functionality you are implementing here because of this. I am one such operator and I certainly want to use what you are implementing :-)

Reply:
My main concern about supporting Annotation at this point is basically about where the namespace-level Annotation comes from.

I am not sure if it is a good idea to get Annotation from arbitrary Route of a namespace, and consider it as the Annotation for the whole namespace as developers (not just operators) have permissions to set the Annotation for each Route, which means it is likely two Routes have different Annotations.

I am happy to see we support Annotation in the wildcard cert after we have a way to configure namespace-level Annotations.

Reply:
It is upto the operator to use an annotation in the template (it is not upto the user) ie the user can set annotation - but it will not be used unless the operator modified the domain template.

The responsibility is therefore on the operator to ensure the right admission controllers or other policy is in place before modifying the domain template (to validate/control the specific annotation on routes).

Reply:
sync'd offline and it seems the issue is that wildcard logic is per ns and not per route. My flawed assumption was that the annotation is consistent across all routes in an ns (which is my use case), but as @ZhiminXiang pointed out other operators may allow different routes to set different values - which will break this.